platforms/nuttx/src/px4/nxp/s32k14x/include/px4_arch/spi_hw_description.h

```diff
@@ -0,0 +1,40 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2020 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+#pragma once
+
+#include "../../../s32k1xx/include/px4_arch/spi_hw_description.h"
+
+constexpr bool validateSPIConfig(const px4_spi_bus_t spi_busses_conf[SPI_BUS_MAX_BUS_ITEMS])
```

Comment:
`#include "../../../s32k1xx/include/px4_arch/spi_hw_description.h"

constexpr bool validateSPIConfig(const px4_spi_bus_t spi_busses_conf[SPI_BUS_MAX_BUS_ITEMS])
{
    const bool nuttx_enabled_spi_buses[] = {
        #ifdef CONFIG_S32K1XX_LPSPI0
                true,
        #else
                false,
        #endif
        #ifdef CONFIG_S32K1XX_LPSPI1
                true,
        #else
                false,
        #endif
        #ifdef CONFIG_S32K1XX_LPSPI2
                true,
        #else
                false,
        #endif
    };

    for (unsigned i = 0; i < sizeof(nuttx_enabled_spi_buses) / sizeof(nuttx_enabled_spi_buses[0]); ++i) {
        bool found_bus = false;

        for (int j = 0; j < SPI_BUS_MAX_BUS_ITEMS; ++j) {
            if (spi_busses_conf[j].bus == (int)i + 1) {
                found_bus = true;
            }
        }

        // Either the bus is enabled in NuttX and configured in spi_busses_conf, or disabled and not configured
        constexpr_assert(found_bus == nuttx_enabled_spi_buses[i], "SPI bus config mismatch (CONFIG_S32K1XX_LPSPIn)");
    }
	return false;
}
`

How's this?

Reply:
I added SPI1 and SPI2 to the hw_description.h file and tested with the pastebin above. I am getting the following output when running `spi bus` in the nsh console: 

nsh> spi bus
 BUS   EXISTS?
Bus 0: NO

Reply:
The is because you (and PX4) are not using the SPI char driver. The errors you are looking for would be compile time. The is what all the constexpr is all about.  Try disabling the CONFIG_S32K1XX_LPSPI0 (comment it out with #) and rebuild.  The build should tell you the validateSPIConfig asserted.

Reply:
Sorry.. I don't completely understand. I disabled CONFIG_S32K1XX_LPSPI0 in the defconfig and rebuilt, and got an error from SPI.cpp: 

../../src/lib/drivers/device/nuttx/SPI.cpp:55:3: error: #error This driver requires CONFIG_SPI_EXCHANGE
   55 | # error This driver requires CONFIG_SPI_EXCHANGE
      |   ^~~~~
compilation terminated due to -Wfatal-errors.

So I tried going into menuconfig and enabling the SPI driver support under Device Drivers -> SPI driver support. In that menu SPI character driver is enabled and SPI exchange is enabled.

I also had to go back in and enable the SPI NuttX app as it dissapeared after I did this. Within that menu I made sure that it had min bus of 0 and max bus of 2.

Still not getting a bus found when running `spi bus` though. So I'm not sure where to go from here. Thank you for your help so far, and sorry for my ignorance. 👍

Reply:
The Validate is suppose to catch  mis configuration CONFIG_S32K1XX_LPSPI0 not enabled but SPI1 used in the boards table. So to test that enable CONFIG_S32K1XX_LPSPI1 and disable CONFIG_S32K1XX_LPSPI0. 


Unless you have configured the SPI character driver The NuttX spi command will not work.  What does `ls /dev` show?

Reply:
Ok - I'm following. If I disable LPSPI0 and enable LPSPI1 in menuconfig, then we get the error:

chip/s32k1xx_lpspi.c: In function 's32k1xx_lpspibus_initialize':
chip/s32k1xx_lpspi.c:1808:29: error: 'PIN_LPSPI1_SCK' undeclared (first use in this function); did you mean 'PIN_LPSPI0_SCK'?
 1808 |           s32k1xx_pinconfig(PIN_LPSPI1_SCK);
      |                             ^~~~~~~~~~~~~~
      |                             PIN_LPSPI0_SCK
chip/s32k1xx_lpspi.c:1808:29: note: each undeclared identifier is reported only once for each function it appears in
chip/s32k1xx_lpspi.c:1809:29: error: 'PIN_LPSPI1_MISO' undeclared (first use in this function); did you mean 'PIN_LPSPI0_MISO'?
 1809 |           s32k1xx_pinconfig(PIN_LPSPI1_MISO);
      |                             ^~~~~~~~~~~~~~~
      |                             PIN_LPSPI0_MISO
chip/s32k1xx_lpspi.c:1810:29: error: 'PIN_LPSPI1_MOSI' undeclared (first use in this function); did you mean 'PIN_LPSPI0_MOSI'?
 1810 |           s32k1xx_pinconfig(PIN_LPSPI1_MOSI);
      |                             ^~~~~~~~~~~~~~~
      |                             PIN_LPSPI0_MOSI
make[1]: *** [Makefile:138: s32k1xx_lpspi.o] Error 1

PIN_LPSPI1_xxxx are not defined because the UCANS32K146 board only has no internal SPI buses and only one external SPI bus. PIN_LPSPI0_xxxx are defined in `boards/nxp/ucans32k146/include/board.h`.

`boards/nxp/ucans32k146/src/spi.cpp` should be board-specific, since it is within the scope of the UCANS32K146, correct? So the spi.cpp file shouldn't check for LPSPI1 or 2?

But the src files in the `platforms/nuttx/src/px4/nxp/s32k1xx/ (and s32k14x)` folders are chip specific, so those need to include all of the available SPI buses on S32K. 

Now I'm understanding the constexpr function, it basically just checks to see if the configured buses in spi.cpp match the CONFIG_S32K1XX_LPSPIn buses. But we never assert because when you enable LPSPI1 or 2, s32k1xx_lpspi.c complains. So what do we do here? Would we want to make some LPSPI1 and 2 defines even though they aren't used?

Also, when I run ls /dev/, there isn't anything that indicates SPI is registered as a char driver. But I enabled it in menuconfig, so why isn't it there?

Thanks.